"permission denied" when proxmox_virtual_environment_file

**Describe the bug**
I get `failed to create file /var/lib/vz/snippets/user-data.yml: permission denied` when creating a snippet from a local file using the enviroment_file resource. 

This very might possibly be a permissions issue -- I can't find much information from Proxmox on what permissions I need to create a snippet. However I followed the instructions to set up my terraform user has the ability to sudo without a password, and I can make the file with sudo. So it seems like the SSH agent isn't using sudo? 

**To Reproduce**
Steps to reproduce the behavior:

1. Create a user "terraform" with "Administrator" on '/'
2. Run `terraform apply` with the following resource:

```hcl

resource "proxmox_virtual_environment_file" "cloud_config" {
  content_type = "snippets"
  datastore_id = "local"
  node_name    = "olympus"

  source_file {
    path = "cloud-init/user-data.yml"
  }
}

```

**Expected behavior**
I would expect the file to be created, or the error to tell me what I need to do to create the file. 

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Additional context**
Add any other context about the problem here.

- Single or clustered Proxmox: single
- Provider version (ideally it should be the latest version): 0.45.1
- Terraform version: 1.7.1
- OS (where you run Terraform from): Linux code-server 4.19.0-26-amd64 #1 SMP Debian 4.19.304-1 (2024-01-09) x86_64 GNU/Linux
- Debug logs (`TF_LOG=DEBUG terraform apply`):
```
2024-01-28T20:24:43.473Z [DEBUG] provider.terraform-provider-proxmox_v0.45.1: SSH connection established: @module=proxmox tf_provider_addr=registry.terraform.io/bpg/proxmox tf_resource_type=proxmox_virtual_environment_file @caller=github.com/bpg/terraform-provider-proxmox/proxmox/ssh/client.go:304 host=192.168.0.100:22 tf_mux_provider=tf5to6server.v5tov6Server tf_req_id=49d75294-ebb5-5684-d33b-5df59da70f22 tf_rpc=ApplyResourceChange user=terraform timestamp=2024-01-28T20:24:43.473Z
2024-01-28T20:24:43.825Z [ERROR] provider.terraform-provider-proxmox_v0.45.1: Response contains error diagnostic: diagnostic_detail= diagnostic_severity=ERROR tf_proto_version=6.4 tf_provider_addr=registry.terraform.io/bpg/proxmox tf_req_id=49d75294-ebb5-5684-d33b-5df59da70f22 @caller=github.com/hashicorp/terraform-plugin-go@v0.21.0/tfprotov6/internal/diag/diagnostics.go:62 @module=sdk.proto diagnostic_summary="failed to create file /var/lib/vz/snippets/user-data.yml: permission denied" tf_resource_type=proxmox_virtual_environment_file tf_rpc=ApplyResourceChange timestamp=2024-01-28T20:24:43.825Z
2024-01-28T20:24:43.879Z [ERROR] vertex "proxmox_virtual_environment_file.cloud_config" error: failed to create file /var/lib/vz/snippets/user-data.yml: permission denied
╷
│ Error: failed to create file /var/lib/vz/snippets/user-data.yml: permission denied
│ 
│   with proxmox_virtual_environment_file.cloud_config,
│   on files.tf line 12, in resource "proxmox_virtual_environment_file" "cloud_config":
│   12: resource "proxmox_virtual_environment_file" "cloud_config" {
│ 
╵
2024-01-28T20:24:43.926Z [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2024-01-28T20:24:43.926Z [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/bpg/proxmox/0.45.1/linux_amd64/terraform-provider-proxmox_v0.45.1 pid=11329
2024-01-28T20:24:43.927Z [DEBUG] provider: plugin exited
```

--- a/.vscode/settings.json
+++ b/.vscode/settings.json
@@ -1,9 +1,11 @@
 {
     "git.alwaysSignOff": true,
     "cSpell.words": [
+        "capi",
         "CLRF",
         "iothread",
         "keyctl",
+        "nolint",
         "proxmoxtf",
         "qcow",
         "rootfs",
@@ -13,7 +15,8 @@
         "virtio",
         "VLANID",
         "vmbr",
-        "VMID"
+        "VMID",
+        "vztmpl"
     ],
     "go.lintTool": "golangci-lint",
     "go.lintFlags": [

--- a/docs/index.md
+++ b/docs/index.md
@@ -160,12 +160,12 @@ You can configure the `sudo` privilege for the user via the command line on the 
     sudo visudo
     ```
 
-  Add the following line to the end of the file:
+  Add the following lines to the end of the file:
 
     ```sh
     terraform ALL=(root) NOPASSWD: /sbin/pvesm
     terraform ALL=(root) NOPASSWD: /sbin/qm
-    terraform ALL=(root) NOPASSWD: /usr/bin/echo tfpve
+    terraform ALL=(root) NOPASSWD: /usr/bin/mv /tmp/tfpve/* /var/lib/vz/*
     ```
 
   Save the file and exit.
@@ -179,10 +179,10 @@ You can configure the `sudo` privilege for the user via the command line on the 
 - Test the SSH connection and password-less `sudo`:
   
     ```sh
-    ssh terraform@<target-node> sudo echo tfpve
+    ssh terraform@<target-node> sudo pvesm apiinfo 
     ```
 
-  You should be able to connect to the target node and see the output `tfpve` on the screen without being prompted for your password.
+  You should be able to connect to the target node and see the output containing `APIVER <number>` on the screen without being prompted for your password.
 
 ### Node IP address used for SSH connection
 

--- a/proxmoxtf/resource/file.go
+++ b/proxmoxtf/resource/file.go
@@ -601,7 +601,7 @@ func fileCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag
 			),
 		})
 		if err != nil {
-			if matches, _ := regexp.Match(`cannot move .* Permission denied`, []byte(err.Error())); matches {
+			if matches, e := regexp.MatchString(`cannot move .* Permission denied`, err.Error()); e == nil && matches {
 				return diag.FromErr(newErrSSHUserNoPermission(capi.SSH().Username()))
 			}
 

--- a/proxmoxtf/resource/file.go
+++ b/proxmoxtf/resource/file.go
@@ -17,10 +17,12 @@ import (
 	"net/url"
 	"os"
 	"path/filepath"
+	"regexp"
 	"sort"
 	"strings"
 	"time"
 
+	"github.com/google/uuid"
 	"github.com/hashicorp/go-cty/cty"
 	"github.com/hashicorp/terraform-plugin-log/tflog"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
@@ -538,7 +540,12 @@ func fileCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag
 	switch *contentType {
 	case "iso", "vztmpl":
 		uploadTimeout := d.Get(mkResourceVirtualEnvironmentFileTimeoutUpload).(int)
+
 		_, err = capi.Node(nodeName).Storage(datastoreID).APIUpload(ctx, request, uploadTimeout, config.TempDir())
+		if err != nil {
+			diags = append(diags, diag.FromErr(err)...)
+			return diags
+		}
 	default:
 		// For all other content types, we need to upload the file to the node's
 		// datastore using SFTP.
@@ -565,14 +572,43 @@ func fileCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag
 			}...)
 		}
 
-		remoteFileDir := *datastore.Path
+		// the temp directory is used to store the file on the node before moving it to the datastore
+		// will be created if it does not exist
+		tempFileDir := fmt.Sprintf("/tmp/tfpve/%s", uuid.NewString())
 
-		err = capi.SSH().NodeUpload(ctx, nodeName, remoteFileDir, request)
-	}
+		err = capi.SSH().NodeUpload(ctx, nodeName, tempFileDir, request)
+		if err != nil {
+			diags = append(diags, diag.FromErr(err)...)
+			return diags
+		}
 
-	if err != nil {
-		diags = append(diags, diag.FromErr(err)...)
-		return diags
+		// handle the case where the file is uploaded to a subdirectory of the datastore
+		srcDir := tempFileDir
+		dstDir := *datastore.Path
+
+		if request.ContentType != "" {
+			srcDir = tempFileDir + "/" + request.ContentType
+			dstDir = *datastore.Path + "/" + request.ContentType
+		}
+
+		_, err := capi.SSH().ExecuteNodeCommands(ctx, nodeName, []string{
+			// the `mv`` command should be scoped to the specific directories in sudoers!
+			fmt.Sprintf(`%s; try_sudo "mv %s/%s %s/%s" && rm -rf %s`,
+				trySudo,
+				srcDir, *fileName,
+				dstDir, *fileName,
+				tempFileDir,
+			),
+		})
+		if err != nil {
+			if matches, _ := regexp.Match(`cannot move .* Permission denied`, []byte(err.Error())); matches {
+				return diag.FromErr(newErrSSHUserNoPermission(capi.SSH().Username()))
+			}
+
+			diags = append(diags, diag.Errorf("error moving file: %s", err.Error())...)
+
+			return diags
+		}
 	}
 
 	volID, di := fileGetVolumeID(d)

--- a/proxmoxtf/resource/file.go
+++ b/proxmoxtf/resource/file.go
@@ -592,12 +592,13 @@ func fileCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag
 		}
 
 		_, err := capi.SSH().ExecuteNodeCommands(ctx, nodeName, []string{
-			// the `mv`` command should be scoped to the specific directories in sudoers!
-			fmt.Sprintf(`%s; try_sudo "mv %s/%s %s/%s" && rm -rf %s`,
+			// the `mv` command should be scoped to the specific directories in sudoers!
+			fmt.Sprintf(`%s; try_sudo "mv %s/%s %s/%s" && rm %s/%s && rmdir -p %s`,
 				trySudo,
 				srcDir, *fileName,
 				dstDir, *fileName,
-				tempFileDir,
+				srcDir, *fileName,
+				srcDir,
 			),
 		})
 		if err != nil {

--- a/proxmoxtf/resource/sudo.go
+++ b/proxmoxtf/resource/sudo.go
@@ -1,0 +1,15 @@
+package resource
+
+import (
+	"fmt"
+)
+
+const (
+	trySudo = `try_sudo(){ if [ $(sudo -n pvesm apiinfo 2>&1 | grep "APIVER" | wc -l) -gt 0 ]; then sudo $1; else $1; fi }`
+)
+
+func newErrSSHUserNoPermission(username string) error {
+	return fmt.Errorf("the SSH user '%s' does not have required permissions. "+
+		"Make sure 'sudo' is installed and the user is configured in sudoers file. "+
+		"Refer to the documentation for more details.", username)
+}

--- a/proxmoxtf/resource/sudo.go
+++ b/proxmoxtf/resource/sudo.go
@@ -11,5 +11,5 @@ const (
 func newErrSSHUserNoPermission(username string) error {
 	return fmt.Errorf("the SSH user '%s' does not have required permissions. "+
 		"Make sure 'sudo' is installed and the user is configured in sudoers file. "+
-		"Refer to the documentation for more details.", username)
+		"Refer to the documentation for more details", username)
 }

--- a/proxmoxtf/resource/vm.go
+++ b/proxmoxtf/resource/vm.go
@@ -11,6 +11,7 @@ import (
 	"encoding/base64"
 	"errors"
 	"fmt"
+	"regexp"
 	"sort"
 	"strconv"
 	"strings"
@@ -2974,7 +2975,7 @@ func vmCreateCustomDisks(ctx context.Context, d *schema.ResourceData, m interfac
 		commands = append(
 			commands,
 			`set -e`,
-			`try_sudo(){ if [ $(sudo -n echo tfpve 2>&1 | grep "tfpve" | wc -l) -gt 0 ]; then sudo $1; else $1; fi }`,
+			trySudo,
 			fmt.Sprintf(`file_id="%s"`, fileID),
 			fmt.Sprintf(`file_format="%s"`, fileFormat),
 			fmt.Sprintf(`datastore_id_target="%s"`, datastoreID),
@@ -3007,9 +3008,8 @@ func vmCreateCustomDisks(ctx context.Context, d *schema.ResourceData, m interfac
 
 		out, err := api.SSH().ExecuteNodeCommands(ctx, nodeName, commands)
 		if err != nil {
-			if strings.Contains(err.Error(), "pvesm: not found") {
-				return diag.Errorf("The configured SSH user '%s' does not have the required permissions to import disks. "+
-					"Make sure `sudo` is installated and the user is a member of sudoers.", api.SSH().Username())
+			if matches, _ := regexp.Match(`pvesm: .* not found`, out); matches {
+				return diag.FromErr(newErrSSHUserNoPermission(api.SSH().Username()))
 			}
 
 			return diag.FromErr(err)

--- a/proxmoxtf/resource/vm.go
+++ b/proxmoxtf/resource/vm.go
@@ -3008,7 +3008,7 @@ func vmCreateCustomDisks(ctx context.Context, d *schema.ResourceData, m interfac
 
 		out, err := api.SSH().ExecuteNodeCommands(ctx, nodeName, commands)
 		if err != nil {
-			if matches, _ := regexp.Match(`pvesm: .* not found`, out); matches {
+			if matches, e := regexp.Match(`pvesm: .* not found`, out); e == nil && matches {
 				return diag.FromErr(newErrSSHUserNoPermission(api.SSH().Username()))
 			}
 

